lib/mix/lib/mix/utils.ex

```diff
@@ -29,17 +29,36 @@ defmodule Mix.Utils do
 
   """
   def mix_config do
-    mix_home_xdg_lookup("XDG_CONFIG_HOME")
+    mix_home_lookup(["MIX_HOME", {:system, :user_config}])
   end
 
-  defp mix_home_xdg_lookup(xdg) do
-    case {System.get_env("MIX_HOME"), System.get_env(xdg)} do
-      {directory, _} when is_binary(directory) -> directory
-      {nil, directory} when is_binary(directory) -> Path.join(directory, "mix")
-      {nil, nil} -> Path.expand("~/.mix")
+  defp mix_home_lookup([]), do: Path.expand("~/.mix")
+
+  defp mix_home_lookup([{:system, type} | rest]) do
+    case xdg?() do
+      true -> xdg_dir(type)
+      _ -> mix_home_lookup(rest)
     end
   end
 
+  defp mix_home_lookup([{env, suffix} | rest]) when is_binary(env) and is_binary(suffix) do
+    case System.get_env(env) do
+      nil -> mix_home_lookup(rest)
+      dir -> Path.join(dir, suffix)
+    end
+  end
+
+  defp mix_home_lookup([env | rest]) when is_binary(env) do
+    mix_home_lookup([{env, ""} | rest])
+  end
```

Comment:
This clause and the previous seems they could be merged? I.e. the suffix is always empty, so there is no reason to have both.

Reply:
It is relict of the previous approach where I manually checked `XDG_*` variables instead of using `:filename`, so yes, it could be merged.